cub/cub/device/dispatch/tuning/tuning_unique_by_key.cuh

```diff
@@ -344,364 +320,273 @@ template <class KeyT,
+  template <int NOMINAL_4B_ITEMS_PER_THREAD, int THREADS>
+  struct DefaultPolicy
   {
-    static constexpr int INPUT_SIZE = sizeof(KeyT);
-    enum
-    {
-      NOMINAL_4B_ITEMS_PER_THREAD = 9,
-      ITEMS_PER_THREAD            = Nominal4BItemsToItems<KeyT>(NOMINAL_4B_ITEMS_PER_THREAD),
-    };
-
+    static constexpr int nominal_4b_items_per_thread = NOMINAL_4B_ITEMS_PER_THREAD;
+    static constexpr int ITEMS_PER_THREAD            = Nominal4BItemsToItems<KeyT>(nominal_4b_items_per_thread);
     using UniqueByKeyPolicyT =
-      AgentUniqueByKeyPolicy<128,
+      AgentUniqueByKeyPolicy<THREADS,
                              ITEMS_PER_THREAD,
                              cub::BLOCK_LOAD_WARP_TRANSPOSE,
                              cub::LOAD_LDG,
                              cub::BLOCK_SCAN_WARP_SCANS,
                              detail::default_delay_constructor_t<int>>;
   };
 
-  struct DefaultTuning
-  {
-    static constexpr int INPUT_SIZE = sizeof(KeyT);
-    enum
-    {
-      NOMINAL_4B_ITEMS_PER_THREAD = 11,
-      ITEMS_PER_THREAD            = Nominal4BItemsToItems<KeyT>(NOMINAL_4B_ITEMS_PER_THREAD),
-    };
+  struct Policy350
+      : DefaultPolicy<9, 128>
+      , ChainedPolicy<350, Policy350, Policy350>
+  {};
 
-    using UniqueByKeyPolicyT =
-      AgentUniqueByKeyPolicy<64,
-                             ITEMS_PER_THREAD,
-                             cub::BLOCK_LOAD_WARP_TRANSPOSE,
-                             cub::LOAD_LDG,
-                             cub::BLOCK_SCAN_WARP_SCANS,
-                             detail::default_delay_constructor_t<int>>;
-  };
+  // Use values from tuning if a specialization exists, otherwise pick the default
+  template <typename Tuning>
+  static auto select_agent_policy(int)
+    -> AgentUniqueByKeyPolicy<Tuning::threads,
+                              Tuning::items,
+                              Tuning::load_algorithm,
+                              Tuning::load_modifier,
+                              BLOCK_SCAN_WARP_SCANS,
+                              typename Tuning::delay_constructor>;
+  template <typename Tuning>
+  static auto select_agent_policy(long) -> typename DefaultPolicy<Tuning::items, Tuning::threads>::UniqueByKeyPolicyT;
 
-  // SM520
   struct Policy520
-      : DefaultTuning
+      : DefaultPolicy<11, 64>
       , ChainedPolicy<520, Policy520, Policy350>
   {};
 
-  /// SM80
   struct Policy800 : ChainedPolicy<800, Policy800, Policy520>
   {
-    using tuning = detail::unique_by_key::sm80_tuning<KeyT, ValueT>;
-
-    using UniqueByKeyPolicyT =
-      AgentUniqueByKeyPolicy<tuning::threads,
-                             tuning::items,
-                             tuning::load_algorithm,
-                             tuning::load_modifier,
-                             BLOCK_SCAN_WARP_SCANS,
-                             typename tuning::delay_constructor>;
+    using UniqueByKeyPolicyT = decltype(select_agent_policy<sm80_tuning<KeyT, ValueT>>(0));
```

Comment:
@bernhardmgruber Here it fails with

>[build] /home/coder/cccl/lib/cmake/cub/../../../cub/cub/device/dispatch/tuning/tuning_unique_by_key.cuh(573): error: no instance of overloaded function "cub::CUB_200800_SM_600_700_800::detail::unique_by_key::policy_hub<KeyT, ValueT>::select_agent_policy [with KeyT=uint8_t, ValueT=c2h::custom_type_t<c2h::equal_comparable_t>]" matches the argument list
[build]             argument types are: (int)
[build]       using UniqueByKeyPolicyT = decltype(select_agent_policy<sm80_tuning<KeyT, ValueT>>(0));
[build]                                           ^